src/utils/page-functions.js

```diff
@@ -0,0 +1,202 @@
+import { createHash } from 'crypto'
+import path from 'path'
+import webpack from 'webpack'
+import { camelCase } from '../utils'
+import {
+  readFile,
+  exists,
+  stats,
+  requireResolve
+} from './fs'
+
+let packageName
+let nodeModulesPath
+
+/* export async function listPageFunctionSources(fnPath) {
+  if (!fnPath) {
+    fnPath = path.join(__dirname, 'page-functions')
+  }
+
+  const files = await readDir(fnPath, { encoding: 'utf8' })
+  return files.filter(file => file.endsWith('.js'))
+} */
+
+export function createPageFunctions(page, sourceFiles) {
+  const pageFunctions = {}
+
+  if (!sourceFiles) {
+    return pageFunctions
+  }
+
+  const babelPresets = page.getBabelPresetOptions()
+
+  for (let file of sourceFiles) {
+    let fnName
+    if (Array.isArray(file)) {
+      [file, fnName] = file
+    } else {
+      fnName = camelCase(path.basename(file, '.js'))
+    }
+
+    pageFunctions[fnName] = async (...args) => {
+      const body = await getPageFunctionBody(fnName, file, babelPresets)
+
+      return page.runAsyncScript({
+        args: [],
+        body: `${body} return (pageFn.${fnName} || pageFn.default).apply(null, arguments)`
```

Comment:
Any reason calling apply with null instead of undefined?

Reply:
Not really, just did it because mdn does that too: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Function/apply